### PR TITLE
Fix/unsafe cast

### PR DESCRIPTION
**Pull request by iK4tsu**
```md
Changes:
 * Simplified and optimized all functions overall
 * Utilize `isAlphaNum` from std.ascii instead of using hardcoded variables
 * Removed unsafe cast from `size_t`  to `uint`
 * Declared all functions as `@safe`
 * Used contracts for extra checks safety in debug mode
 * Splitted unittests by each function
 * Used `enum` to declare the constant variables, you can also use immutable instead.
```

There's something I don't understand in the last function `usernameRemoveDuplicateSymbols`. Why are you removing sequential *symbols* but most importantly why are you removing sequential *symbols* which differ from each other? If a user explicitly types `Foo-.-Bar` I assume it's because he wants the username that way. With the implementation, you currently have the same user can login with `Foo-Bar` which can lead to major confusion to the user. Moreover, this is unsafe for your users as well as they might think they have a safer user name by using those combinations when in fact they are just trimmed internally, which also means brute-force attacks will be easier to do. If you want to remove sequential *symbols* I'd recommend only remove the ones which are equal.

I changed the username sequence because you have these constraints and as such it does not make sense to search for a username in the database with an invalid username because such a username will never be in the database.

With these changes, you can make these functions templates and use them at compile time :). For your current use-case, this isn't necessary it's just a little tip. Here's an example of what it could look like:
```d
template containsSpecialCharacters(string username)
{
    import std.algorithm : canFind, each;
    import std.ascii : isAlphaNum;
    import std.typecons : No, Yes;

    enum containsSpecialCharacters = username.each!((ref c) => (!isAlphaNum(c) && !symbols.canFind(c)) ? No.each : Yes.each) == No.each;
}

template startsWithSymbol(string username)
{
    import std.ascii : isAlphaNum;
    enum usernameStartsWithSymbol = !isAlphaNum(username[0]);
}

template endsWithSymbol(string username)
{
    import std.ascii : isAlphaNum;
    enum usernameStartsWithSymbol = !isAlphaNum(username[$ - 1]);
}

template isUsername(string username)
{
    enum isUsername = !containsSpecialCharacters!username && !startsWithSymbol!username && !endsWithSymbol!username;
}

template removeSequencialSymbols(string username)
    if (isUsername!username)
{
    import std.algorithm : reverse, uniq;
    import std.array : array;
    import std.ascii : isAlphaNum;
    import std.conv : to;
    import std.range : retro;

    enum removeSequencialSymbols = username.retro.uniq!((a, b) => !isAlphaNum(a) && !isAlphaNum(b)).array.reverse.to!string;
}
```

Then just call them with:
```d
assert(isValidUsername!"Valid_User");
assert(removeSequencialSymbols!"Valid__--User" == "Valid-User");
```

If in the future, you decide to change the behavior of the `usernameRemoveDuplicateSymbols` function to remove only the sequential *symbols* which are equal you can do it this way:
```d
public string usernameRemoveDuplicateSymbols(in string username) @safe
{
    import std.algorithm : uniq;
    import std.ascii : isAlphaNum;
    import std.conv : to;

    return username.uniq!((a, b) => !std.ascii.isAlphaNum(a) && a == b).to!string;
}
```

### CODE REVIEW

**Comment by CodeMyst** [comment]
The reason I guess for the symbol removal is to reduce spammy usernames. And the issue you mentioned that the user doesn't know the username is trimmed isn't a big issue since the username is only cosmetic, it servers no purpose since you don't use it to log in.

**Comment by FFY00** [comment]
Won't this just result in impersonation? And what does this removing of spammy usernames mean? You will still get the spam... And this would only flag a portion of the spam usernames anyway.
I find that a better approach would be flagging the accounts as potential spam, a configuration switch could also be added to enable admins to require manual approval for flagged accounts.
But overall I find this could be easily mitigated by introducing per IP registration limits or a captcha. The current approach doesn't actually reduce spam or send out warnings. And, my biggest issue, it has way too many false positives, I can think of a few people I know with sequences in their usernames. IMO this will just end up generating confusion between users.
Anyway, that's just my two cents :blush:

File: source/pastemyst/util/username.d [hunk 1]
```diff
@@ -2,112 +2,153 @@ module pastemyst.util.username;
 
 import std.string;
 
-private const(string) alpha = "abcdefghijklmnopqrstuvwxyz";
-private const(string) nums = "0123456789";
-private const(string) alphanums = alpha ~ nums;
-private const(string) symbols = "-_.";
```
Comment (ljmf00) [inline source/pastemyst/util/username.d:8]: Just an hint here @CodeMyst . The compiler converts this value type to immutable values. If you do const(T) the compiler will do immutable(const (T)). You should use immutable or enum for compile-time constant values

**Comment by iK4tsu** [comment]
I wouldn't go as far as impersonation, but I can't say fully I agree with your argument either @CodeMyst. From my point of view, this action won't stop users from creating *spammy* usernames. This will only prevent a very low percentage. In my opinion, this username `Foo__--Foo` is less *spammy* as `FooooooooooooooooooooFooooooooooooooooooooo` and this `Foo_______-------F......O-----O` is just as *spammy* as last. See where I'm getting at? One is allowed the other two aren't. The username might only be cosmetic but it is the user's choice of how he wants his username to look like. Making these partial restrictions and behaviors might lead to the user's confusion and frustration. Using a flagging system, setting up some ground rules for usernames, and letting the admins take care of that problem manually would be a nice path to follow as @FFY00 suggested. All this being said, this is just an observation :) . The changes I've made in this PR do not change the behavior you currently have in the slightest, and putting this aside I hope you enjoy them ^ ^.

**Comment by Dmunch04** [comment]
Isn't a flagging system and admins going through usernames a bit much? Just for spammy usernames lol. Or did I get it wrong

**Comment by iK4tsu** [comment]
The issue here is about the current way of handling this problem. I just don't think this is the right way to approach it. It might be a bit much just for usernames, but you could extend such a system to no only usernames but to pastes as well. Potential harming links, harassing, etc. There are other ways to do it of course this was just an idea.

**Comment by Dmunch04** [comment]
Hmm yes. I think @CodeMyst is also working on some solution. I have this quote from the Discord:
> yeah that is an issue i will figure out how to deal with as the problems come. issues with people reporting posts for example or spammers. i wont be deleting posts because of personal data i dont think (your fault for posting it), or for piracy links.
so yeah, will figure it out as it comes

**Comment by CodeMyst** [comment]
ok yeah, you have all changed my mind on this!

if i dont have any strict rules on what can be posted on the sites, why should i have strict rules on usernames, and not even proper rules but just reducing amount of special characters, which won't make things any better in reality.

i am thinking of a system for reporting pastes and users, but for the v2 release there wont be anything like that yet.

so i guess that now the username symbol rules should be removed completely.

**Comment by iK4tsu** [comment]
How do you want to proceed? Should we make this change in this PR or make a separate PR tackling the issue? Personally I think the latter is better as it could lead to other fixes in relation to this issue.

**Comment by CodeMyst** [comment]
yeah sure, you can create a separate PR. im gonna close this PR then since all changes arent relevant anymore I think

**Comment by iK4tsu** [comment]
Only the function `usernameRemoveDuplicateSymbols` is applied to this issue. You can still have the other functions as ground rules for the usernames. Not containing special characters and not starting and/or ending with a symbol are valid ground rules since these apply to everyone and do not have any *side effects* internally.

**Comment by CodeMyst** [comment]
ah right, im stupid lol

**Pull request merged**